File: ssrf_explorer/ssrf/detector.py
"""Classify probe responses into SSRF severity buckets."""
from __future__ import annotations

import re
from typing import Dict, List, Optional

META_MARKERS = (
    "ami-id",
    "instance-identity",
    "iam/security-credentials",
    "computeMetadata",
    "Metadata-Flavor",
    "access_token",
    "securityCredentials",
)
BANNER_MARKERS = (
    "SSH-2.0",
    "SSH-1.99",
    "220 ",
    "Redis",
    "MEMCACHED",
    "HTTP/1.0 4",
    "Server: gunicorn",
)
FILE_MARKERS = (
    "root:x:0:0",
    "[extensions]",
    "[fonts]",
    "for 16-bit app support",
)


def classify(
    payload_category: str,
    status: Optional[int],
    body_excerpt: str,
    elapsed_ms: int,
    baseline_ms: Optional[int] = None,
    oob_hit: bool = False,
) -> tuple[str, List[str]]:
    signals: List[str] = []
    body = body_excerpt or ""

    if any(m in body for m in META_MARKERS):
        signals.append("cloud-metadata-content")
    if any(m in body for m in BANNER_MARKERS):
        signals.append("service-banner")
    if any(m in body for m in FILE_MARKERS):
        signals.append("local-file-content")
    if status and 200 <= status < 400 and payload_category in ("loopback", "private"):
        signals.append(f"internal-reachable-{status}")
    if baseline_ms is not None and elapsed_ms - baseline_ms > 1500:
        signals.append(f"timing-delta-{elapsed_ms - baseline_ms}ms")
    if oob_hit:
        signals.append("oob-callback")

    if "oob-callback" in signals or "cloud-metadata-content" in signals or "local-file-content" in signals:
        severity = "Confirmed"
    elif "service-banner" in signals or any(
        s.startswith("internal-reachable-") for s in signals
    ):
        severity = "Likely"
    elif signals:
        severity = "Possible"
    else:
        severity = "None"
    return severity, signals

File: server/services/detector.js
// Classify a probe response into an SSRF severity bucket.

const META = ['ami-id', 'instance-identity', 'iam/security-credentials', 'computeMetadata', 'Metadata-Flavor', 'securityCredentials'];
const BANNER = ['SSH-2.0', 'SSH-1.99', '220 ', 'Redis', 'MEMCACHED', 'Server: gunicorn'];
const FILE = ['root:x:0:0', '[extensions]', '[fonts]', 'for 16-bit app support'];

// Near-identity check that tolerates small per-request differences (CSRF
// tokens, timestamps, request IDs) but still catches "the response is the
// same as the authenticated baseline" — which is the clearest proof that the
// server ignored our payload and there's no SSRF.
function bodiesEffectivelyEqual(a, b) {
  if (!a || !b) return false;
  if (a === b) return true;
  if (Math.abs(a.length - b.length) / Math.max(a.length, b.length) > 0.05) return false;
  const strip = (s) => s
    .replace(/[0-9a-f]{16,}/gi, '')                 // long hex tokens
    .replace(/\d{4}-\d{2}-\d{2}T[\d:.+-Z]+/g, '')   // ISO timestamps
    .replace(/\b\d{10,}\b/g, '')                    // epoch-like integers
    .replace(/\s+/g, ' ')
    .trim();
  return strip(a) === strip(b);
}

function classify({ category, status, body, elapsedMs, baselineMs, baselineBody, oobHit }) {
  const signals = [];
  const b = body || '';
  const bb = baselineBody || '';
  const unchanged = bodiesEffectivelyEqual(b, bb);

  if (META.some((m) => b.includes(m))) signals.push('cloud-metadata-content');
  if (BANNER.some((m) => b.includes(m))) signals.push('service-banner');
  if (FILE.some((m) => b.includes(m))) signals.push('local-file-content');

  // Only call an endpoint "internal-reachable" when its response actually
  // CHANGED vs the baseline. A 200 with an unchanged body means the server
  // returned its normal answer and silently dropped our payload — no SSRF.
  if (status && status >= 200 && status < 400 && (category === 'loopback' || category === 'private')) {
    if (unchanged) {
      signals.push('response-unchanged-vs-baseline');
    } else {
      signals.push(`internal-reachable-${status}-body-changed`);
    }
  }

  if (baselineMs != null && elapsedMs - baselineMs > 1500 && !unchanged) {
    signals.push(`timing-delta-${elapsedMs - baselineMs}ms`);
  }

  if (oobHit) signals.push('oob-callback');

  let severity;
  if (
    signals.includes('oob-callback') ||
    signals.includes('cloud-metadata-content') ||
    signals.includes('local-file-content')
  ) {
    severity = 'Confirmed';
  } else if (signals.includes('service-banner')) {
    severity = 'Likely';
  } else if (signals.some((s) => s.startsWith('internal-reachable-'))) {
    // "Body changed + status 2xx on a loopback payload" is a real but weak
    // signal — could also be a WAF page or a generic error handler. Report
    // as Possible, let the operator drill in. Combined with a timing delta
    // we raise to Likely.
    severity = signals.some((s) => s.startsWith('timing-delta-')) ? 'Likely' : 'Possible';
  } else if (signals.length && !signals.includes('response-unchanged-vs-baseline')) {
    severity = 'Possible';
  } else {
    severity = 'None';
  }
  return { severity, signals };
}

module.exports = { classify };
